File: xcloak-agent-desktop/agent/suppression_test.go
package agent

import (
	"encoding/json"
	"fmt"
	"testing"
)

// TestSuppressionRule_URLPatterns verifies correct URL construction for
// suppression rule operations.
func TestSuppressionRule_URLPatterns(t *testing.T) {
	cases := []struct {
		got  string
		want string
	}{
		{fmt.Sprintf("/api/suppression/rules/%d", 9), "/api/suppression/rules/9"},
		{fmt.Sprintf("/api/suppression/rules/%d/toggle", 9), "/api/suppression/rules/9/toggle"},
	}
	for _, c := range cases {
		if c.got != c.want {
			t.Errorf("URL = %q, want %q", c.got, c.want)
		}
	}
}

// TestSuppressionRulePayload_Fields verifies that a suppression rule record
// includes the fields needed for the frontend to display and manage rules.
func TestSuppressionRulePayload_Fields(t *testing.T) {
	raw := `{"id":1,"name":"Block Nmap","rule_name":"nmap_scan","window_minutes":60,"enabled":true,"match_count":3}`
	var rule map[string]any
	if err := json.Unmarshal([]byte(raw), &rule); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, key := range []string{"id", "name", "rule_name", "window_minutes", "enabled"} {
		if _, ok := rule[key]; !ok {
			t.Errorf("suppression rule payload missing field %q", key)
		}
	}
}

// TestToggleSuppressionRule_BodyShape verifies that the PATCH toggle body
// matches what the backend expects: `{"enabled": bool}`.
func TestToggleSuppressionRule_BodyShape(t *testing.T) {
	body := map[string]any{"enabled": false}
	b, _ := json.Marshal(body)
	var decoded map[string]any
	json.Unmarshal(b, &decoded)
	if _, ok := decoded["enabled"]; !ok {
		t.Error("toggle body must include 'enabled' key")
	}
}

File: xcloak-agent-mobile/test/suppression_test.dart
// Tests for suppression rule API shapes consumed by the mobile agent.
// The mobile agent may display suppressed alert states; suppression rules
// are read from the backend to decide whether to surface notifications.
//
// Run with: flutter test

import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Suppression Rules (mobile API shape)', () {
    test('suppression rule record includes required fields', () {
      final rule = {
        'id': 1,
        'name': 'Block Nmap',
        'rule_name': 'nmap_scan',
        'window_minutes': 60,
        'enabled': true,
        'match_count': 3,
      };

      for (final key in ['id', 'name', 'rule_name', 'window_minutes', 'enabled']) {
        expect(rule.containsKey(key), isTrue,
            reason: 'suppression rule must include "$key" for the suppression UI');
      }
    });

    test('toggle body uses enabled bool field', () {
      final body = {'enabled': false};
      expect(body['enabled'], isA<bool>(),
          reason: 'toggle PATCH body must use enabled as bool per backend contract');
    });

    test('suppression rule URL patterns are correct', () {
      String ruleUrl(int id) => '/api/suppression/rules/$id';
      String toggleUrl(int id) => '/api/suppression/rules/$id/toggle';

      expect(ruleUrl(9), equals('/api/suppression/rules/9'));
      expect(toggleUrl(9), equals('/api/suppression/rules/9/toggle'));
    });
  });
}
